pkg/stub/handler.go

```diff
@@ -247,3 +255,179 @@ func (h *Handler) ensureRouterDeleted(ci *ingressv1alpha1.ClusterIngress) error
+			co.Namespace, co.Name, err)
+	}
+}
+
+// getOperatorState gets and returns the resources necessary to compute the
+// operator's current state.
+func (h *Handler) getOperatorState() (*corev1.Namespace, *ingressv1alpha1.ClusterIngressList, *appsv1.DaemonSetList, error) {
+	ns, err := h.manifestFactory.RouterNamespace()
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf(
+			"error building router namespace: %v", err)
+	}
+
+	if err := sdk.Get(ns); err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil, nil, nil
+		}
+
+		return nil, nil, nil, fmt.Errorf(
+			"error getting Namespace %s: %v", ns.Name, err)
+	}
+
+	ingressList := &ingressv1alpha1.ClusterIngressList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ClusterIngress",
+			APIVersion: "ingress.openshift.io/v1alpha1",
+		},
+	}
+	err = sdk.List(h.namespace, ingressList,
+		sdk.WithListOptions(&metav1.ListOptions{}))
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf(
+			"failed to list ClusterIngresses: %v", err)
+	}
+
+	daemonsetList := &appsv1.DaemonSetList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "DaemonSet",
+			APIVersion: "extensions/v1beta1",
+		},
+	}
+	err = sdk.List(h.namespace, daemonsetList,
+		sdk.WithListOptions(&metav1.ListOptions{}))
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf(
+			"failed to list DaemonSets: %v", err)
+	}
+
+	return ns, ingressList, daemonsetList, nil
+}
+
+// setStatusConditions sets the specified ClusterOperator's status using the
+// provided resources to compute the operator's current state.
+func setStatusConditions(co *osv1.ClusterOperator, ns *corev1.Namespace, ingressList *ingressv1alpha1.ClusterIngressList, daemonsetList *appsv1.DaemonSetList) {
+	failingCondition := &osv1.ClusterOperatorStatusCondition{
+		Type:   osv1.OperatorFailing,
+		Status: osv1.ConditionUnknown,
+	}
+	if ns == nil {
+		failingCondition.Status = osv1.ConditionTrue
+		failingCondition.Reason = "NoNamespace"
+		failingCondition.Message = "router namespace does not exist"
+	} else {
+		failingCondition.Status = osv1.ConditionFalse
+	}
+	clusteroperator.SetStatusCondition(co, failingCondition)
+
+	progressingCondition := &osv1.ClusterOperatorStatusCondition{
+		Type:   osv1.OperatorProgressing,
+		Status: osv1.ConditionUnknown,
+	}
+	if ingressList != nil && daemonsetList != nil {
+		numIngresses := len(ingressList.Items)
+		numDaemonsets := len(daemonsetList.Items)
+		if numIngresses == numDaemonsets {
+			progressingCondition.Status = osv1.ConditionFalse
+		} else {
+			progressingCondition.Status = osv1.ConditionTrue
+			progressingCondition.Reason = "Reconciling"
+			progressingCondition.Message = fmt.Sprintf(
+				"have %d ingresses, want %d",
+				numDaemonsets, numIngresses)
+		}
+	}
+	clusteroperator.SetStatusCondition(co, progressingCondition)
+
+	availableCondition := &osv1.ClusterOperatorStatusCondition{
+		Type:   osv1.OperatorAvailable,
+		Status: osv1.ConditionUnknown,
+	}
+	if ingressList != nil && daemonsetList != nil {
```

Comment:
I'm not sure this correlation is correct, the DaemonSet name is a function of (not equal to) the ClusterIngress name... also, I think you have to compare availability of daemonsets relative to their own status: e.g.:

* for each ingress
  * get daemonset by ingress
    * let fullyAvailable = (daemonset.desired == daemonset.ready)

I think a more natural way to do this sort of stuff would be to build up some indexed caches from the k8s stdlib. For example:

* DaemonSet cache with index:
  * by ClusterIngress.Name where .Status.Desired == .Status.Ready

Then given a ClusterIngress you can do an O(1) lookup to see if there exists a ready DaemonSet associated with the ClusterIngress.

Reply:
Maybe it would be better to start small and slightly incorrect and build up to unblock deployment? For example, if the count of daemonsets matches the count of ingresses, at least mark us as "Progressing" and not failed, and we can refine the algorithms in followups using strategies like in 
https://github.com/openshift/cluster-ingress-operator/pull/47/commits/5a46915b18c90fd4883fb95d7672607fef25c4e8#r226467331.

Would that help?

Reply:
> I'm not sure this correlation is correct, the DaemonSet name is a function of (not equal to) the ClusterIngress name...

Fixed (I think).  Thanks!

> also, I think you have to compare availability of daemonsets relative to their own status: e.g.:
> 
> * for each ingress
>   * get daemonset by ingress
>     * let fullyAvailable = (daemonset.desired == daemonset.ready)

What is `daemonset.desired`, and what is `daemonset.ready`? I chose to use `ds.Status.NumberAvailable` based on my reading the docstrings for `OperatorAvailable` and `DaemonSetStatus` and a literal interpretation of "available".  If my interpretation is not correct, we should specify what we mean by "available".

> I think a more natural way to do this sort of stuff would be to build up some indexed caches from the k8s stdlib.

I've been trying to get by using operator-sdk.  Presumably its client API will grow richer, so I don't want to move away from it prematurely.

Reply:
Available is probably right, I didn't look closely at the docs for those fields.

Regarding indexers, no need to go outside the operator-sdk. [Check this out](https://github.com/kubernetes/client-go/blob/master/informers/apps/v1/daemonset.go#L50). Given just a `kubernetes.Interface` (easily available at our entrypoint in `main.go`) you can create any number of indexes to pass along to the handler.

There are a lot of examples to help you prototype in unit tests, which should help.

* [shared_informer_test.go](https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer_test.go)
* [index_test.go](https://github.com/kubernetes/client-go/blob/master/tools/cache/index_test.go)
* [controller_test.go](https://github.com/kubernetes/client-go/blob/master/tools/cache/controller_test.go) 

Making liberal use of caches/indexes/appropriate data structures is going to be very useful here to avoid an ever-expanding rat's nest of imperative analysis code.